Split rendering driver project setting into `rendering_method` and `rendering/*/driver`

clayjohn opened the pull request:
Closes: https://github.com/godotengine/godot-proposals/issues/4261

This also exposes OpenGL to the project creation menu and re-exposes the renderer switch button in the top right hand corner of the editor.

Things to consider during review:
1. Organization of the project settings. I organized the project settings as suggested in https://github.com/godotengine/godot-proposals/issues/4261, but it feels a little awkward.
2. Using enums instead of strings for renderer name and rendering driver. In 3.x we had a VideoDriver enum which we used throughout the engine instead of exposing strings directly. I think we should consider having a RenderingDriver enum and a RendererName enum which we pass around to avoid doing risky string operations all over the engine.

I blindly made some changes to the iOS and Android platforms, would be nice if someone from each team can confirm that the changes look okay.

Screenshots of new project settings:

![Screenshot from 2022-09-12 14-13-11](https://user-images.githubusercontent.com/16521339/189759852-80d01618-d15e-4fde-8e68-4dabfcb640cf.png)

_when using gl_compatibility_
![Screenshot from 2022-09-12 14-13-05](https://user-images.githubusercontent.com/16521339/189759854-d3b0d96e-2a13-4a8a-95ac-603289df0f7d.png)

_when using rendering device based renderers_
![Screenshot from 2022-09-12 14-12-57](https://user-images.githubusercontent.com/16521339/189759857-381b73b6-22e2-4bd0-ad86-f2f32612dc61.png)

clayjohn committed:
Split rendering driver project setting into renderer_name and rendering_driver. To differentiate between a driver (e.g. Vulkan or D3D12) and a renderer (e.g. clustered or mobile renderer).

Geometror commented:
Great work! However, I wonder if it's worth it to reintroduce the the renderer switch button in the menu bar. This is something which will be changed very rarely during development for that it takes quite some space.

clayjohn commented:
> Great work! However, I wonder if it's worth it to reintroduce the the renderer switch button in the menu bar. This is something which will be changed very rarely during development for that it takes quite some space.

I was thinking the same thing. Especially since the new renderer names are much longer than GLES3/GLES2 (as they were in 3.x). I was thinking I would enable it here and then let the UI/UX people fix/remove it as they see fit.

aaronfranke commented:
Why hide the rendering API from the user? Why call GLES3 just called "Compatibility"? I think it would make sense to let users know that it's GLES3. Same for the Vulkan ones, I don't see a reason to hide that it's Vulkan.

HeadClot commented:
> Why hide the rendering API from the user? Why is GLES3 just called "Compatibility"? I think it would make sense to let users know that it's GLES3. Same for the Vulkan ones, I don't see a reason to hide that it's Vulkan.

I am with aaron on this. I think that hiding the api name from user will lead to confusion. My 2 cents.

Calinou commented:
> Why hide the rendering API from the user? Why is GLES3 just called "Compatibility"? I think it would make sense to let users know that it's GLES3. Same for the Vulkan ones, I don't see a reason to hide that it's Vulkan.

In 4.0, different drivers can be used for a same rendering backend. The Clustered and Mobile backends will be usable both with Vulkan and Direct3D 12 depending on the target hardware and user preference.

The Compatibility backend will be usable both with OpenGL directly and with ANGLE (for GPUs where OpenGL support is problematic).

aaronfranke commented:
@Calinou I disagree, it's extremely relevant, not an implementation detail.

* Graphics APIs are something that every game developer has some knowledge about, and even many gamers know about the existence of Vulkan, OpenGL, and DirectX.

* The choice of graphics API is *directly* related to the hardware it can run on. If somebody knows what specific devices they want to support, they can look up which APIs those devices support.

m4gr3d commented on platform/android/export/export_plugin.cpp:
```diff
@@ -1670,14 +1670,7 @@ Vector<String> EditorExportPlatformAndroid::get_enabled_abis(const Ref<EditorExp
-	if (driver == "vulkan") {
-		r_features->push_back("etc2");
-	}
+	r_features->push_back("etc2");
```
Shouldn't we have `etc` for the `compatibility/GLES3` driver?

clayjohn replied:
OpenGL ES 3.0, OpenGL 3,3, and WebGL 2 all support ETC2 which is a superset of ETC. So essentially we can treat everything as if it is ETC2. Internally we no longer check for ETC support, just ETC2

m4gr3d commented:
> I blindly made some changes to the iOS and Android platforms, would be nice if someone from each team can confirm that the changes look okay.

@clayjohn The Android changes looks good.

clayjohn commented:
> Why hide the rendering API from the user? Why is GLES3 just called "Compatibility"? I think it would make sense to let users know that it's GLES3. Same for the Vulkan ones, I don't see a reason to hide that it's Vulkan.

Its not hidden. OpenGL3 is the driver, not the renderer. The choice of driver is still presented to the user. On top of that, Users have never been able to explicitly set the Rendering API. The GLES3 driver mostly confirmed to OpenGL ES 3.0 like the name implies, but on web it is limited to the set of features in the WebGL specification and on Desktop it is limited by OpenGL 3.3. This PR lets us actually expose the selection of rendering API to the user.

BastiaanOlij commented:
Sorry @clayjohn , bit late to the discussion and about to potentially throw a monkey wrench into it :)

Looking to the (near) future and the wish to make the engine more extendable with more implementations, either full renderers like the GLES3 one (possibly a raytracing backend that was discussed), different drivers such as our DirectX backend that works with the RD renderer as a replacement/alternative to Vulkan, or additional RD renderers such as the planned deferred renderer, having these selections hardcoded the way that we do will become a nightmare.

I wonder if we need some way to register the available options and make the selection on the projection creation window and dropdown selection more generic?

YuriSizov commented:
I don't see harm in printing the supported driver/API in the hint text for each option. That's definitely not something that would stand in the way of this PR, just a bit about how we present the information to the user.

Options and how they separate target platforms make sense to me. The names may need to be bikeshed, but it's not too critical.

clayjohn commented:
Notes from September 9th meeting:
1. Change compatibility to gl_compatibility
2. Have two codepaths one for compatibility and one for rendering_device
3. Rename clustered to forward+

clayjohn commented:
Updated to:
1. remove GL_Compatibility renderer from project selection (will expose in another PR)
2. remove renderer selection from upper-right hand corner of editor (needs more discussion how this should be exposed)
3. rename "clustered" renderer to "forward_plus"
4. rename "compatibility" renderer to "gl_compatibility"

Should now be ready for review by @reduz

clayjohn commented:
@aaronfranke @HeadClot I have removed the change to the project manager and I have updated the top post to highlight how this PR allows users to directly choose their graphics API for selected rather than before where it was fixed based on renderer choice. 

Please let me know if you find this more suitable. To be perfectly frank, I don't fully understand your opposition to this PR in the first. We can't name the mobile renderer "Vulkan Mobile" because it may be running over Metal, WebGPU, D3D12, _or_ Vulkan. One thing we could do is expose renderer choice and driver choice in the project manager so users can explicitly choose the renderer they want to us and the graphics API that they want to target.

aaronfranke commented:
@clayjohn Good points. With Metal, as far as I know it's using MoltenVK, which is Vulkan on top of Metal, but I guess maybe this isn't the case for D3D12 and WebGPU. Anyway, what about something like this?

![Screen Shot 2022-09-12 at 5 29 45 PM](https://user-images.githubusercontent.com/1646875/189770313-d581101e-60ba-4c9c-b0ea-452cc25b93b0.png)

clayjohn commented:
> @clayjohn Good points.With Metal, as far as I know it's using MoltenVK, which is Vulkan on top of Metal, but I guess maybe this isn't the case for D3D12 and WebGPU. Anyway, the menus you presented make a lot of sense to me, so the driver name is still shown, just not in the project manager. How about something like this?
> 
> ![Screen Shot 2022-09-12 at 5 29 45 PM](https://user-images.githubusercontent.com/1646875/189770313-d581101e-60ba-4c9c-b0ea-452cc25b93b0.png)

That seems fine to me. Although to be clear, I have reverted the project manager changes in this PR and I won't make further changes in this PR. Those changes will have to be made in another PR. This PR is just about updating project settings now. I understand that reduz wants to make a proposal for the Renderer/Driver selection in the Project Manager.

reduz commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1941,6 +1955,44 @@
 		<member name="rendering/reflections/sky_reflections/texture_array_reflections.mobile" type="bool" setter="" getter="" default="false">
 			Lower-end override for [member rendering/reflections/sky_reflections/texture_array_reflections] on mobile devices, due to performance concerns or driver support.
 		</member>
+		<member name="rendering/renderer/renderer_name" type="String" setter="" getter="" default="&quot;forward_plus&quot;">
```
I think we should use "rendering_method" rather than "renderer_name". This should make it less ambiguous with "rendering_driver". Maybe we shuld eventually also rename the RenderScene class to RenderMethod given it will now become pretty much that.

reduz commented:
I think its good, but I feel "renderer_name" can be open to confusion with the driver, IMO we should rename both the setting and the class to "rendering method", to better differentiate. I am also wondering whether "rendering driver" should be renamed to "rendering device driver".

reduz commented:
Looks good to me, we may need to clean up main at some point, but can be done at another time. Need agreement from @BastiaanOlij about renaming the renderer scene class to rendering method.

BastiaanOlij reviewed (approved):
Looks good to me, I'm happy with the rename :)

clayjohn commented:
Rebased, so should be ready to merge! @akien-mga 

One thing I shouldv'e highlighted in the original post is that projects made in an earlier 4.0 version will have the project setting warning label appear on older projects in the ProjectManager. The warning can be safely ignored, but we will want to flag it in the next couple of blog posts:
![image](https://user-images.githubusercontent.com/16521339/190502583-05ce626b-309f-4702-bc30-0c28d8424bcd.png)

akien-mga commented:
> One thing I shouldv'e highlighted in the original post is that projects made in an earlier 4.0 version will have the project setting warning label appear on older projects in the ProjectManager. The warning can be safely ignored, but we will want to flag it in the next couple of blog posts:

To avoid confusion, maybe we can keep the old names temporarily for compatibility, and add some code to actually remove them from `project.godot` when opening in beta 2+?

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1947,6 +1961,44 @@
 		</member>
+		<member name="rendering/renderer/rendering_method" type="String" setter="" getter="" default="&quot;forward_plus&quot;">
+			Sets the renderer that will be used by the project. Options are:
+				[b]Clustered[/b]: High-end renderer designed for Desktop devices. Has a higher base overhead, but scales well with complex scenes. Not suitable for older devices or mobile.
+				[b]Mobile[/b]: Modern renderer designed for mobile devices. Has a lower base overhead than Clustered, but does not scale as well to large scenes with many elements.
+				[b]Compatibility[/b]: Low-end renderer designed for older devices. Based on the limitations of the OpenGL 3.3/ OpenGL ES 3.0 / WebGL 2 APIs.
```
How does that extra indent render? I don't know we're using this style anywhere in the docs yet. I'd suggest going with either no extra indent, or a `-` list.

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1787,6 +1780,27 @@
+			LinuxBSD override for [member rendering/gl_compatibility/driver].
+		</member>
+		<member name="rendering/gl_compatibility/driver.macos" type="String" setter="" getter="" default="&quot;opengl3&quot;">
+			MacOS override for [member rendering/gl_compatibility/driver].
```
```suggestion
			macOS override for [member rendering/gl_compatibility/driver].
```

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1947,6 +1961,44 @@
+			LinuxBSD override for [member rendering/rendering_device/driver].
+		</member>
+		<member name="rendering/rendering_device/driver.macos" type="String" setter="" getter="" default="&quot;vulkan&quot;">
+			MacOS override for [member rendering/rendering_device/driver].
```
```suggestion
			macOS override for [member rendering/rendering_device/driver].
```

akien-mga commented:
Looks fine overall, should be good to merge once the remaining comments are solved.

clayjohn commented:
> Looks fine overall, should be good to merge once the remaining comments are solved.

Thanks for checking! Pushed a fix with your suggested changes

aaronfranke commented on core/config/project_settings.cpp:
```diff
@@ -101,10 +101,15 @@ const PackedStringArray ProjectSettings::_get_supported_features() {
+
+#ifdef VULKAN_ENABLED
+	features.append("forward_plus");
+	features.append("mobile");
```
These names are designed to be user-friendly since they can be displayed to the user, so they should likely be "Forward Plus" etc.

aaronfranke commented on editor/project_manager.cpp:
```diff
@@ -473,16 +473,17 @@ class ProjectDialog : public ConfirmationDialog {
+					initial_settings["rendering/renderer/rendering_method"] = renderer_type;
+
+					if (renderer_type == "forward_plus" || renderer_type == "mobile") {
+						project_features.push_back("Vulkan");
```
There is a mismatch between here and project_settings.cpp. As-is, this will add "Vulkan" to the list of features when creating a new project, but this isn't on the list of supported features.

clayjohn commented:
@akien-mga @aaronfranke Thank you both for reviewing! All of your comments should be addressed now

akien-mga commented:
Thanks!
